### ISSUE TITLE
Game Wizard 32 Tutor causes DOSBox-X to crash

### ISSUE THREAD

**Issue by rderooy** [joncampbell123/dosbox-x#1434]
**Describe the bug**
I tried running Game Wizard 32 Pro 3.0a in DOSBox-X. And while at first glance the main program seems to work, running the included GWTUTOR.EXE causes dosbox-x to crash (core dump).

GW32 is a TSR program to make your own cheats for DOS games.

**To Reproduce**
Install Game Wizard from disk and run GWTUTOR.EXE. Note! Any change to DOSBox version or settings seems to require a reinstall from disk due to some kind of checksum copy protection that it does. There is also a shareware version of GW32 that I did not try.

**Expected behavior**
The tutor program to do its thing

**Screenshots**

**Environment (please complete the following information):**
Fedora 31
dosbox-x-0.83.0-0.el8.x86_64

**Additional context**
Here are some of the messages from the terminal. First I ran GW.EXE and then GWTUTOR.EXE
LOG:  110290139 ERROR DOSMISC:DOS:INT 2F Unhandled call AX=4680
LOG: DOSBox has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox's options.
LOG:  110726982 ERROR MISC:DOS:Set File Attributes for C:\GW32PRO\X\REGISTER.DOC not supported
LOG:  110727003 ERROR MISC:DOS:Set File Attributes for C:\GW32PRO\X\GW.OV2 not supported
LOG:  110727021 ERROR MISC:DOS:Set File Attributes for C:\GW32PRO\X\GW.BOS not supported
LOG:  110727041 ERROR MISC:DOS:Set File Attributes for C:\GW32PRO\X\GW.OV3 not supported
LOG:  110727046 ERROR MISC:EMS:Call FF not supported
LOG:  219914849 ERROR DOSMISC:DOS:INT 2F Unhandled call AX=4680
LOG: Restart by keyboard controller requested
LOG: CMOS Shutdown byte 0x0a says to jump to reset vector 0a2e:05d6
terminate called after throwing an instance of 'int'
Aborted (core dumped)

**Comment by Allofich** [comment]
> LOG: 219914849 ERROR DOSMISC:DOS:INT 2F Unhandled call AX=4680

This, at least, should not be the issue. It's a check for whether Windows is running and MS-DOS would be ignoring it, too.

**Comment by Allofich** [comment]
I can confirm the crash.

The crash happens if you start GWTUTOR.EXE with the default "auto" core, or with Dynamic Core. If you select Normal, Full or Single for the core, the crash doesn't happen. (Side issue: In the GUI, Normal appears as selected by default (has a checkmark) when using auto core, which is misleading)

The log errors mentioned above about file attributes, etc. seem to be unrelated. The full log I get when running with the crash with the latest code (so `INT 2F Unhandled call AX=4680` no longer appears) is

```
EXEC:Execute gwtutor.EXE 0
FILES:file open command 0 file gwtutor.EXE
DEBUG MISC:DOS_AllocateMemory(blocks=0x0045) = 0x02e3-0x0327
DEBUG MISC:DOS_AllocateMemory(blocks=0x9cd6) = 0x0329-0x9ffe
DOSBox has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox's options.
Restart by keyboard controller requested
CMOS Shutdown byte 0x0a says to jump to reset vector 06b4:05d6
```

**Comment by rderooy** [comment]
Thanks for the update Allofich.

**Comment by Allofich** [comment]
For me, when the crash happens it goes nuts at address `06B4:0FE1`.

The function it's in looks like:

```
06B4:0FBE
mov al,CF
call 00000E87
mov bx,ss:[1038]
lgdt word ss:[bx+2C]
mov bx,ss:[103A]
lidt word ss:[bx+2C]
pop ax
push word ss:[104C]
push ax
mov ax,0001
lmsw ax 06B4:0FE1 (<-Crash)
retf
```

Seems this is related to switching to Protected Mode.
https://stackoverflow.com/questions/30932302/i-am-confusing-some-assembly-code-about-enable-pe-within-boot-setup-s-file-in-linux-0-11

I'm sure @joncampbell123 will know more about it.

**Comment by rderooy** [comment]
There is a new thread on vogons about this: https://www.vogons.org/viewtopic.php?f=33&t=73018
Apparently you need to set core=normal and ems=false, but even then there are still problems with the programs memory state save function, at least with regular DOSBox. Will try what happens with DOSBox-X with those settings when I have a moment.

**Comment by rderooy** [comment]
Quick update using 0.83.20.

Starting with the default dynamic core:
```
LOG: DOSBox-X has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox-X's options.
LOG:    9384644 ERROR MISC:EMS:Call FF not supported
LOG: Restart by keyboard controller requested
LOG: CMOS Shutdown byte 0x0a says to jump to reset vector 0f1d:05d6
LOG: E_Exit: Sorry, CMOS shutdown CPU reset method is not compatible with dynamic core
E_Exit: Sorry, CMOS shutdown CPU reset method is not compatible with dynamic core
```

Starting with `core=normal` the tutor does work. It does not seem necessary to disable EMS, but with EMS enabled you do get an extra error.
```
LOG: DOSBox-X has switched to max cycles, because of the setting: cycles=auto.
If the game runs too fast, try a fixed cycles amount in DOSBox-X's options.
LOG:    7133411 ERROR MISC:EMS:Call FF not supported
LOG: Restart by keyboard controller requested
LOG: CMOS Shutdown byte 0x0a says to jump to reset vector 0f1b:05d6
```

**Comment by grapeli** [comment]
Maybe it depends on the version (copy) you are using. This version runs under dosbox-x and `core=dynamic_x86` provided `cputype` is not newer than `486`.

https://archive.org/details/game-wizard-32-pro-30-a-dfi

**Comment by rderooy** [comment]
Hmm, it is v3.0a Pro. And DOSBox-X exits when running `GWTUTOR.EXE` with an E_Exit regardless if I set `cputype=286`, 386 or 486.

I downloaded the one your linked to, and it also exits the same way as the version I had, regardless of cputype.

**Comment by grapeli** [comment]
https://user-images.githubusercontent.com/452325/144716150-dcdd9ec7-61f7-4ed0-a949-35b592abf939.mp4

**Comment by rderooy** [comment]
yes gw works, but it is the tutor program that does not. ``gwtutor`` as it also says in the title ;-)

**Comment by grapeli** [comment]
I missed it, sorry.

`GWTUTOR.EXE` doesn't actually work with `core=dynamic_x86`, I confirm.

**Comment by grapeli** [comment]
@rderooy 
Generally, this is a **very minor problem**.
Worse, if the main Game Wizard with `core=dynamic_x86` did not work.

Gwtutor is only a demonstration.

**Comment by NebularNerd** [comment]
Just tested this for fun and found the following seems to make this behave for me using the archive link above and installing from the floppy image:   

```
core=normal
output=direct3d
```
The output is semi optional, I found that when using `output=tty` you would get some graphical weirdness if you switch between the `GW.EXE` menu and the back to the `GWTUTOR.EXE` screens as they both use text modes. For regular games there should be no issue.

https://user-images.githubusercontent.com/8470449/189981160-12212606-ccee-474d-9428-f6d49de91859.mp4

And of course the first test has to be making a Tardis ship on Frontier 😁
![frontier_000](https://user-images.githubusercontent.com/8470449/189984575-92a687a0-65ac-47a6-88b5-4a35400a19ff.png)
![frontier_002](https://user-images.githubusercontent.com/8470449/189984598-4b5b1118-aa12-4a1f-b42f-4061b3d3d8a9.png)
![frontier_001](https://user-images.githubusercontent.com/8470449/189984611-f3976d8f-2bd5-4b04-9ac1-ab3870be7ead.png)
Infinite cargo space can be obtained by selling the never ending supply of water, stuff in the biggest Hyperdrive you can lay your hands on (Hilariously fast when you use a lightweight ship with a Class 7 drive) and then insta-jump from system to system.
